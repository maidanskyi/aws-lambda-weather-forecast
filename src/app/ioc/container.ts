import { Container } from 'inversify';

import { TYPES } from '../constants';
import {
  EnvConfigReadable,
  Loggable,
  SecretReadable
} from '../interfaces';
import {
  AwsSecretManagerProvider,
  EnvConfigurationProvider,
  PinoConsoleLoggerProvider
} from '../providers';

export const container = new Container();

container
  .bind<SecretReadable>(TYPES.awsSecretManager)
  .to(AwsSecretManagerProvider);

container
  .bind<EnvConfigReadable>(TYPES.envConfig)
  .to(EnvConfigurationProvider);

container
  .bind<Loggable>(TYPES.pinoConsoleLogger)
  .to(PinoConsoleLoggerProvider);
