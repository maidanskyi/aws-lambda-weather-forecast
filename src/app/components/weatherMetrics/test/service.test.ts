import 'reflect-metadata';
import { container } from '../../../ioc';
import { TYPES } from '../../../constants';
import {
  EnvConfigReadable,
  MetricGettable,
  MetricPublishable,
  MetricReadable
} from '../../../interfaces';

describe('WeatherMetrics service test cases', () => {
  let getMetricsSpy: jest.SpyInstance;
  let response: [string, string];

  beforeAll(async () => {
    // rebind service which is used for getting env variables used in the method
    const envConfigService = container.get<EnvConfigReadable>(TYPES.envConfig);

    jest
      .spyOn(envConfigService, 'getAwsRegion')
      .mockImplementation(() => {
        return 'us-east-1';
      });

    container
      .rebind(TYPES.envConfig)
      .toConstantValue(envConfigService);

    // get the service and spy on his method
    const metricsStorageService = container
      .get<MetricPublishable & MetricReadable>(TYPES.dynamoDbStorage);

    getMetricsSpy = jest
      .spyOn(metricsStorageService, 'getMetrics')
      .mockImplementation(async () => {
        return {
          successResponses: 12,
          failureResponses: 1,
        };
      });

    container
      .rebind(TYPES.dynamoDbStorage)
      .toConstantValue(metricsStorageService);

    // now get the weatherForecastService instance from the container
    // all his dependencies are mocked so we can call the method
    // and check behavior
    const weatherForecastService = container
      .get<MetricGettable>(TYPES.weatherMetricsService);

    response = await weatherForecastService.getMetrics();
  });

  it('should call spy method', () => {
    expect(getMetricsSpy).toHaveBeenCalled();
  });

  it('should return metrics with mocked values specified above', () => {
    expect(response).toMatchSnapshot();
  });
});
