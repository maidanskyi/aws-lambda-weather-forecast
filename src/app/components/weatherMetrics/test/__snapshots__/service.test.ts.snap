// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherMetrics service test cases should return metrics with mocked values specified above 1`] = `
Array [
  "text/plain; version=0.0.4; charset=utf-8",
  "# HELP requestCounter Requests statistics
# TYPE requestCounter counter
requestCounter{successRequests=\\"12\\",failureRequests=\\"1\\"} 1
",
]
`;
